### ISSUE TITLE
Humidifier - Integrate function for Level 1,2,3

### ISSUE THREAD

**Issue by Kohle81** [OpenWonderLabs/homebridge-switchbot-openapi#85]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

hello, temporarily I also bought the humidifier and it works perfectly!  I like to have the humidifier always on level 3, this cannot yet be controlled via homebridge / Homekit, right?  If the humidifier is manually set to level 1, it will only be heard on the pre-set level via an automation.  would be great if you could do this.

**Comment by donavanbecker** [comment]
@Kohle81 how would you expect this to be displayed?

**Comment by Kohle81** [comment]
Thanks for fast answer. It could be a dimmable lamp. 1-33 % is for level 1, 34-66 % is level 2 and 64-100 % level 3.......

**Comment by donavanbecker** [comment]
what I will do is add an option for a min step value that can be set in the config. and then if you see that to 33 then it will go from 33/66/99... when you set it.

**Comment by Kohle81** [comment]
This is great! Thanks

**Comment by donavanbecker** [comment]
@Kohle81 okay got time to work on this, can you test the latest beta?

Beta: `v2.1.1-beta.0`

**Comment by Kohle81** [comment]
I have installed beta.3 and it works.

If I use the light lamp in Homekit, the level switches from 1 to 2 or 3......

It works fine!

When I switches in the switchbot app the level from 2to 3 the light lamp doesn‘t actuql status...... It keeps to 48 %

**Comment by SwitchBot-Wonderlabs** [comment]
> I have installed beta.3 and it works. If I use the light lamp in Homekit, the level switches from 1 to 2 or 3...... It works fine! When I switches in the switchbot app the level from 2to 3 the light lamp doesn‘t actuql status...... It keeps to 48 %
> […](#)
> Am 01.02.2021 um 04:49 schrieb Donavan Becker ***@***.***>: ﻿ @Kohle81 okay got time to work on this, can you test the latest beta? Beta: v2.1.1-beta.0 — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

Hey, @Kohle81 , please try again and feedback with the app.

**Comment by Kohle81** [comment]
Hi, I am on 2.1.1-beta.0 now.

If the humidifier is switches on/off manually, there is no reaction in Homekit. The state would not be actualized....

**Comment by Kohle81** [comment]
Hmmh, I am now back to the actual version.

There is the same, if I switch on the humidifier manually, there is noch actual state in Homekit. The device is on and Homekit shows it at off. In the switchbot app the state is on.

Yesterday it worked perfectly. Could it be, that a restart of homebridge deleted something?

**Comment by donavanbecker** [comment]
@Kohle81 No I would try installing the beta again. I am running it and tried the minSteps and it is working as expected.

**Comment by Kohle81** [comment]
I tried the beta again and now it works actually. I have checked my config, there is a refresh rate= 120....... so I am sorry, „mea culpa“

**Comment by Kohle81** [comment]
I am on beta.3 and the state of the humidifier  will actually after 3 min.

But I noticed, that the State of the humidifier is Level 1 (look Switchbot App) and Homekit shows 100 %. 

![72B99DD6-0746-47B9-A000-CFEA9EB3CDFA](https://user-images.githubusercontent.com/33038709/106501699-bc91bc80-64c3-11eb-97d6-9601127e0bff.jpeg)
![Uploading A1734609-53AE-4A73-A98C-08C651A2D870.jpeg…]()

**Comment by donavanbecker** [comment]
@Kohle81 can you try  Beta:`v2.1.1-beta.7`?

**Comment by Kohle81** [comment]
hi, i'm now on beta 7. Unfortunately, manual switching does not update the status in Homekit correctly.  the humidifier starts up, but shows 100%, although the humidifier is at low level

**Comment by donavanbecker** [comment]
> hi, i'm now on beta 7. Unfortunately, manual switching does not update the status in Homekit correctly.  the humidifier starts up, but shows 100%, although the humidifier is at low level

I don't know what to tell you, because this is working perfectly for me on my humidifier.

**Comment by Kohle81** [comment]
Ok, thanks for your answer.

I will have a look tonight. Perhaps I would delete the cached accessories and restart homebridge....

**Comment by Kohle81** [comment]
Hi,
I am now correct at beta.7, there was a failure with „sudo npm install „, I tried with „unsafe perm“ and now beta7 is installed.

But now Humidifier isn‘t in Homekit and I have this report in log:

Feb 03 19:21:06 raspberrypi homebridge[6163]: [03/02/2021, 19:21:06] Homebridge is running on port 51811.
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Total SwitchBot Devices Found: 7
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Total IR Devices Found: 1
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Bett Nicki hoch Bot
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Bett Nicki hoch Bot DeviceID: C42F040FCCD2
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Bett Manu runter Bot
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Bett Manu runter Bot DeviceID: C9A1CF72C9FC
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Bett Manu hoch Bot
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Bett Manu hoch Bot DeviceID: CBA415F0A044
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Switchbot Luftqualität Meter
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Switchbot Luftqualität Meter DeviceID: D1FFDE76E505
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Luftbefeuchter Humidifier
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Luftbefeuchter Humidifier DeviceID: FCF5C43E85EE
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered a Hub MiniFeb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Bett Nicki runter Bot
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Bett Nicki runter Bot DeviceID: FEDAC1EDB7BD
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Discovered Haier TV DIY TV
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] [SwitchBot] Adding new accessory: Haier TV DIY TV DeviceID: 02-202009290754-73988876Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] Haier TV DIY TV is running on port 38079.
Feb 03 19:21:07 raspberrypi homebridge[6163]: [03/02/2021, 19:21:07] Please add [Haier TV DIY TV] manually in Home app. Setup Code: 031-45-154
Feb 03 19:21:08 raspberrypi homebridge[6163]: (node:6163) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'set_minStep' of undefined
Feb 03 19:21:08 raspberrypi homebridge[6163]:     at new Humidifier (/usr/local/lib/node_modules/homebridge-switchbot-openapi/src/Devices/Humidifier.ts:101:60)
Feb 03 19:21:08 raspberrypi homebridge[6163]:     at SwitchBotPlatform.createHumidifier (/usr/local/lib/node_modules/homebridge-switchbot-openapi/src/platform.ts:361:7)
Feb 03 19:21:08 raspberrypi homebridge[6163]:     at SwitchBotPlatform.discoverDevices (/usr/local/lib/node_modules/homebridge-switchbot-openapi/src/platform.ts:188:18)
Feb 03 19:21:08 raspberrypi homebridge[6163]:     at processTicksAndRejections (internal/process/task_queues.js:93:5)
Feb 03 19:21:08 raspberrypi homebridge[6163]: (Use `node --trace-warnings ...` to show where the warning was created)
Feb 03 19:21:08 raspberrypi homebridge[6163]: (node:6163) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
Feb 03 19:21:08 raspberrypi homebridge[6163]: (node:6163) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Feb 03 19:21:09 raspberrypi homebridge[6163]: [03/02/2021, 19:21:09] [SwitchBot] Humidifier - Failed to update status of Luftbefeuchter "Cannot read property 'updateCharacteristic' of undefined" undefined               

Here is my config:

"platforms": [{
		"platform": "SwitchBot",
		"name": "SwitchBot",
		"credentials": {
			"openToken": "xxxxxxxxxxxyyyyyyyyy",
			"notice": "Keep your token a secret!"
		},
		"options": {
			"refreshRate": 120,
			"bot": {
				"device_press": [
					"C42F040FCCD2",
					"C9A1CF72C9FC",
					"CBA415F0A044",
					"FD6A96EC2808"
				]
			}
		},
		"devicediscovery": false
	}]

**Comment by donavanbecker** [comment]
@Kohle81 can you update to Beta: `v2.1.1-beta.14`

**Comment by Kohle81** [comment]
I am now on beta 14, the humidifier is displayed again in homekit.  There are no errors in the log.  but the status does not change in homekit.  the humidifier was on level 1 and after the plugin update and homebridge restart it took a few minutes.  then the humidifier was on, but level at 100%

**Comment by donavanbecker** [comment]
> I am now on beta 14, the humidifier is displayed again in homekit.  There are no errors in the log.  but the status does not change in homekit.  the humidifier was on level 1 and after the plugin update and homebridge restart it took a few minutes.  then the humidifier was on, but level at 100%
> 
> > Am 04.02.2021 um 07:16 schrieb Donavan Becker <notifications@github.com>:
> > 
> > ﻿
> > @Kohle81 can you update to Beta: v2.1.1-beta.14?
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> 

Can you clear activity on just your humidifier accessory from homebridge UI? You don't have to clear all of  your homebridge accessories.

**Comment by donavanbecker** [comment]
@Kohle81 Once again, not sure what to you tell you? I am displaying exactly what SwitchBot is displaying. Also looking at your config above doesn't look like you have minStep setup for the humidifier...

<a><img src="https://user-images.githubusercontent.com/9875439/106927184-89c01200-66d7-11eb-9aac-07192278acb9.PNG" width="250px"></a><a><img src="https://user-images.githubusercontent.com/9875439/106927173-888ee500-66d7-11eb-8824-dd6e8b6fbd4f.PNG" width="250px"></a>

**Comment by Kohle81** [comment]
Can you show me please, how to configure config.json?

**Comment by donavanbecker** [comment]
@Kohle81 Easiest way will be to use Homebridge UI

**Comment by Kohle81** [comment]
I understand that.  I try to get the setting for the config from others.  I am grateful for your willingness and help.  I am used to the fact that the config is available on github ........ but I don't want to appear ungrateful ....

**Comment by Kohle81** [comment]
Hi,

thanks for showing your config, I didn't realized this over the info e-mail from GitHub.......

I have changed my config like:

{
	"bridge": {
		"name": "Homebridge-switchbot",
		"username": "CC:22:3D:E3:CE:11",
		"port": 51811,
		"pin": "031-45-154",
		"manufacturer": "@Paul Schmidt",
		"model": "Homebridge",
		"serialNumber": "0.4.45"
	},
	"plugins": [
		"homebridge-switchbot-openapi"
	],

	"accessories": [],

	"platforms": [{
		"platform": "SwitchBot",
		"name": "SwitchBot",
		"credentials": {
			"openToken": "xxxxxxxxxxxxxx",
			"notice": "Keep your token a secret!"
		},
		"options": {
			"bot": {
				"device_press": [
					"C42F040FCCD2",
					"C9A1CF72C9FC",
					"CBA415F0A044",
					"FD6A96EC2808"
				]
			},
			"humidifier": {
				"set_minStep": 20
			}
		},
		"devicediscovery": false
	}]
}



I will answer tomorrow, how it works. Thanks for your great work!

**Comment by Kohle81** [comment]
hello, it's me again.  I am currently making the following observation: if I change the level in the switchbot, the status in homekit is displayed after about 4-5 minutes.  but if I manually operate the switch on the humidifier myself and thus change the level, I currently don't get any changes in homekit - but the status in the switchbot app is current and correct

**Comment by donavanbecker** [comment]
@Kohle81 this is a cloud based api, so homekit has to wait till Homebridge refreshes, the default refresh is set to 300, which would be 5 mins. If you use homebridge UI, it allows you to make this number lower then 300 if you choose.

**Comment by Kohle81** [comment]
Thanks,

I am now back to

"refreshRate": 120,

and now it looks good! 👍

I hope I wasn't too strenuous 😜

**Comment by Kohle81** [comment]
Hi @donavanbecker ,

I have tested the function and synchronisation homekit with minihub works.

Perhaps I have found a bug.

First time, when humidifier is off and then I switch humidifier to on, there is level „low“, par example.

After 2 min. honekit shows humidifier as „on“, but on level 100 %.....

humidifier and switchbot app are showing level „low“.

In this szenario synchronisation doesn‘t work, I am waiting for 5-10 min. tomorrow and tested three times.

If I switch to another level in the switchbot app, then after 2 min. Homekit is synchron and then Level changings are working.
